File: scripts/audit-ai-native-studio-pc2.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-2.0-or-later
"""Reopen and independently audit the PC.2 action derivative."""

import argparse
import hashlib
import json
import math
import os
import sys
from pathlib import Path

import bpy


SENTINELS = (1, 48, 96, 97, 144, 192, 193, 240, 288)


def canonical(value):
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()


def normalize(value):
    if isinstance(value, float) and math.isfinite(value) and value.is_integer():
        return int(value)
    if isinstance(value, list):
        return [normalize(item) for item in value]
    if isinstance(value, dict):
        return {key: normalize(item) for key, item in value.items()}
    return value


def hash_value(value):
    return hashlib.sha256(canonical(normalize(value))).hexdigest()


def sha256_file(path):
    digest = hashlib.sha256()
    with Path(path).open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def valid_self(value, field):
    body = dict(value)
    expected = body.pop(field, None)
    return expected == hash_value(body)


def write_self(path, value, field):
    body = normalize(dict(value))
    body[field] = hash_value(body)
    descriptor = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    try:
        os.write(descriptor, (json.dumps(body, ensure_ascii=False, indent=2) + "\n").encode())
        os.fsync(descriptor)
    finally:
        os.close(descriptor)
    return body


def rounded(values):
    return [round(float(value), 9) for value in values]


def camera_light_state(scene):
    cameras = sorted((obj for obj in bpy.data.objects if obj.type == "CAMERA"), key=lambda item: item.name)
    lights = sorted((obj for obj in bpy.data.objects if obj.type == "LIGHT"), key=lambda item: item.name)
    rows = []
    for frame in SENTINELS:
        scene.frame_set(frame)
        rows.append({
            "frame": frame,
            "cameras": {obj.name: {"matrixWorld": [rounded(row) for row in obj.matrix_world], "lens": round(float(obj.data.lens), 9)} for obj in cameras},
            "lights": {obj.name: {"matrixWorld": [rounded(row) for row in obj.matrix_world], "energy": round(float(obj.data.energy), 9), "color": rounded(obj.data.color)} for obj in lights},
        })
    return rows


def mesh_geometry_state():
    rows = []
    for obj in sorted((item for item in bpy.data.objects if item.type == "MESH"), key=lambda item: item.name):
        mesh = obj.data
        rows.append({"object": obj.name, "data": mesh.name, "parent": obj.parent.name if obj.parent else None, "parentType": obj.parent_type, "parentBone": obj.parent_bone, "vertices": [rounded(vertex.co) for vertex in mesh.vertices], "edges": [list(edge.vertices) for edge in mesh.edges], "polygons": [list(polygon.vertices) for polygon in mesh.polygons], "materials": [slot.material.name if slot.material else None for slot in obj.material_slots], "modifiers": [(modifier.name, modifier.type) for modifier in obj.modifiers]})
    return {"objects": len(bpy.data.objects), "meshes": len(rows), "polygons": sum(len(obj.data.polygons) for obj in bpy.data.objects if obj.type == "MESH"), "canonicalSha256": hash_value(rows)}


def material_state():
    rows = []
    for material in sorted(bpy.data.materials, key=lambda item: item.name):
        nodes = []
        if material.use_nodes and material.node_tree:
            for node in sorted(material.node_tree.nodes, key=lambda item: item.name):
                inputs = []
                for index, socket in enumerate(node.inputs):
                    value = getattr(socket, "default_value", None)
                    if isinstance(value, (float, int)):
                        inputs.append([index, socket.name, round(float(value), 9)])
                    elif value is not None and hasattr(value, "__len__"):
                        try:
                            inputs.append([index, socket.name, rounded(value)])
                        except (TypeError, ValueError):
                            pass
                nodes.append({"name": node.name, "type": node.bl_idname, "inputs": inputs})
        rows.append({"name": material.name, "useNodes": bool(material.use_nodes), "nodes": nodes})
    return {"materials": len(rows), "canonicalSha256": hash_value(rows)}


def shot_state(scene):
    return [{"name": marker.name, "frame": marker.frame, "camera": marker.camera.name if marker.camera else None} for marker in sorted(scene.timeline_markers, key=lambda item: (item.frame, item.name))]


def action_fcurves(action):
    curves = []
    if action is None:
        return curves
    for layer in action.layers:
        for strip in layer.strips:
            for channelbag in strip.channelbags:
                curves.extend(channelbag.fcurves)
    return curves


def curve_for(target, data_path, index):
    action = target.animation_data.action if target.animation_data else None
    matches = [curve for curve in action_fcurves(action) if curve.data_path == data_path and curve.array_index == index]
    if len(matches) != 1:
        raise RuntimeError(f"CURVE_{target.name}_{data_path}_{index}_{len(matches)}")
    return matches[0]


def verify_signal(row):
    target = bpy.data.objects[row["target"]]
    curve = curve_for(target, row["dataPath"], row["arrayIndex"])
    points = {round(float(point.co[0]), 6): float(point.co[1]) for point in curve.keyframe_points}
    actual = []
    for frame, expected in zip(row["frames"], row["values"], strict=True):
        if float(frame) not in points or abs(points[float(frame)] - float(expected)) > 1e-6:
            raise RuntimeError(f"SIGNAL_POINT_{row['target']}_{row['dataPath']}_{frame}")
        actual.append(points[float(frame)])
    peak = max(actual) - min(actual)
    return {"phase": row["phase"], "target": row["target"], "dataPath": row["dataPath"], "arrayIndex": row["arrayIndex"], "frames": row["frames"], "peakToPeak": round(peak, 9), "minimumPeakToPeak": row["minimumPeakToPeak"], "passed": peak + 1e-6 >= row["minimumPeakToPeak"]}


def parse_args():
    argv = sys.argv[sys.argv.index("--") + 1:] if "--" in sys.argv else []
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec", type=Path, required=True)
    parser.add_argument("--evidence-root", type=Path, required=True)
    parser.add_argument("--work-root", type=Path, required=True)
    return parser.parse_args(argv)


args = parse_args()
spec = json.loads(args.spec.read_text(encoding="utf-8"))
build = json.loads((args.evidence_root / "build.json").read_text(encoding="utf-8"))
if not valid_self(spec, "specHash") or not valid_self(build, "buildHash"):
    raise RuntimeError("SELF_HASH")
scene = bpy.context.scene
opened_path = Path(bpy.data.filepath)
if sha256_file(opened_path) != build["derived"]["sha256"]:
    raise RuntimeError("DERIVED_SHA")
if scene.get("bfs_pc2_spec_hash") != spec["specHash"] or scene.get("bfs_pc2_source_sha256") != spec["acceptedPc1Baseline"]["sha256"] or scene.get("bfs_pc2_signal_hash") != hash_value(build["signals"]):
    raise RuntimeError("SCENE_BINDING")

geometry = mesh_geometry_state()
materials = material_state()
protected = camera_light_state(scene)
shots = shot_state(scene)
if geometry != build["geometryAfter"] or geometry != build["geometryBefore"]:
    raise RuntimeError("GEOMETRY")
if materials != build["materialsAfter"] or materials != build["materialsBefore"]:
    raise RuntimeError("MATERIALS")
if protected != build["protectedStateAfter"] or protected != build["protectedStateBefore"]:
    raise RuntimeError("PROTECTED")
if hash_value(protected) != spec["acceptedPc1Baseline"]["cameraLightSentinelsCanonicalSha256"]:
    raise RuntimeError("PROTECTED_BASELINE")
if shots != build["shotsAfter"] or shots != build["shotsBefore"]:
    raise RuntimeError("SHOTS")

signal_checks = [verify_signal(row) for row in build["signals"]]
phase_ids = sorted({row["phase"] for row in signal_checks})
signal_targets = sorted({row["target"] for row in signal_checks})
if not all(row["passed"] for row in signal_checks):
    raise RuntimeError("AMPLITUDE")
if len(phase_ids) != spec["acceptance"]["minimumSemanticTemporalPhases"] or len(signal_targets) < spec["acceptance"]["minimumAnimatedNonCameraTargets"]:
    raise RuntimeError("FLOORS")
if not set(signal_targets).issubset(set(spec["authorizedAnimatedNonCameraTargets"])):
    raise RuntimeError("TARGET_SCOPE")
render_artifacts = sorted(str(path) for root in (args.evidence_root, args.work_root) for path in root.rglob("*") if path.is_file() and path.suffix.lower() in {".exr", ".png", ".jpg", ".jpeg", ".mov", ".mp4"})
if render_artifacts:
    raise RuntimeError("RENDER_ARTIFACTS")

audit = write_self(args.evidence_root / "semantic-audit.json", {
    "schemaVersion": "bfs.pc2SemanticAudit.v0.1",
    "status": "PASS",
    "gate": "PC.2",
    "openedBlend": {"path": str(opened_path), "sha256": sha256_file(opened_path)},
    "geometry": geometry,
    "materials": materials,
    "protectedStateCanonicalSha256": hash_value(protected),
    "shotStateCanonicalSha256": hash_value(shots),
    "phaseIds": phase_ids,
    "signalTargets": signal_targets,
    "signalChecks": signal_checks,
    "independentActionChannels": build["independentActionChannels"],
    "renderArtifacts": render_artifacts,
    "operations": {"BlenderStarts": 1, "renderCalls": 0, "sceneSaves": 0, "networkCalls": 0, "modelCalls": 0, "mouseInteractions": 0},
}, "auditHash")
print("PC2_AUDIT=" + json.dumps({"status": audit["status"], "auditHash": audit["auditHash"]}, sort_keys=True), flush=True)

File: scripts/audit-ai-native-studio-pc2.mjs
#!/usr/bin/env node
import { createHash } from 'node:crypto';
import { open, readFile, readdir, stat } from 'node:fs/promises';
import { dirname, relative, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const repositoryRoot = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const SPEC_URI = 'specs/ai-native-studio-pc2-action-complexity-preregistration.v0.1.json';
const FREEZE_URI = 'specs/ai-native-studio-pc2-tool-freeze.v0.1.json';
const ROOT_URI = 'experiments/ai-native-studio-post-pb7/PC.2-2026-08-31-mac-m2max-attempt-01';
const WORK_ROOT = '/Users/mengyingli/Documents/ChatGPT/BlenderFilmStudio-PostPB7-workspace/PC.2-2026-08-31-mac-m2max-attempt-01';
function sorted(v) { if (Array.isArray(v)) return v.map(sorted); if (v && typeof v === 'object') return Object.fromEntries(Object.entries(v).sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0).map(([k, x]) => [k, sorted(x)])); return v; }
const canonical = v => JSON.stringify(sorted(v)); const sha = v => createHash('sha256').update(v).digest('hex'); const shaFile = async p => sha(await readFile(p));
const without = (v, f) => { const x = structuredClone(v); delete x[f]; return x; }; const validSelf = (v, f) => v?.[f] === sha(canonical(without(v, f)));
function pythonNumber(value) { if (!Number.isFinite(value)) throw new Error('NONFINITE'); if (Number.isInteger(value)) return String(value); return String(value).replace(/e([+-]?)(\d+)$/i, (_, sign, digits) => `e${sign || ''}${digits.padStart(2, '0')}`); }
function pythonCanonical(value) { if (value === null) return 'null'; if (Array.isArray(value)) return `[${value.map(pythonCanonical).join(',')}]`; if (typeof value === 'object') return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${pythonCanonical(value[key])}`).join(',')}}`; if (typeof value === 'number') return pythonNumber(value); return JSON.stringify(value); }
const validPythonSelf = (v, f) => v?.[f] === sha(pythonCanonical(without(v, f)));
const readJson = async p => JSON.parse(await readFile(p, 'utf8'));
async function writeJson(path, body, field) { const value = { ...body, [field]: sha(canonical(body)) }; const h = await open(path, 'wx', 0o600); try { await h.writeFile(`${JSON.stringify(value, null, 2)}\n`); await h.sync(); } finally { await h.close(); } return value; }
async function walk(root) { const out = []; async function visit(path) { for (const e of await readdir(path, { withFileTypes: true })) { const next = resolve(path, e.name); if (e.isSymbolicLink()) throw new Error(`SYMLINK_${next}`); if (e.isDirectory()) await visit(next); else if (e.isFile()) out.push(next); } } await visit(root); return out.sort(); }

async function auditPc2(argv = process.argv.slice(2)) {
  if (argv.length !== 2 || argv[0] !== '--evidence-root' || argv[1] !== ROOT_URI) throw new Error('USAGE');
  const root = resolve(repositoryRoot, ROOT_URI), specPath = resolve(repositoryRoot, SPEC_URI), freezePath = resolve(repositoryRoot, FREEZE_URI);
  const expectedRoster = ['build.json', 'logs', 'processes', 'receipt.json', 'semantic-audit.json'];
  const roster = (await readdir(root)).sort(); if (JSON.stringify(roster) !== JSON.stringify(expectedRoster)) throw new Error(`ROSTER_${roster.join(',')}`);
  const spec = await readJson(specPath), freeze = await readJson(freezePath), build = await readJson(resolve(root, 'build.json')), semantic = await readJson(resolve(root, 'semantic-audit.json')), receipt = await readJson(resolve(root, 'receipt.json'));
  const processPaths = ['processes/01-build.json', 'processes/02-semantic-audit.json']; const processes = await Promise.all(processPaths.map(path => readJson(resolve(root, path))));
  const checks = []; const gate = (id, pass, detail) => checks.push({ id, pass: Boolean(pass), detail });
  gate('SELF_HASHES', validSelf(spec, 'specHash') && validSelf(freeze, 'specHash') && validPythonSelf(build, 'buildHash') && validPythonSelf(semantic, 'auditHash') && validPythonSelf(receipt, 'receiptHash') && processes.every(row => validPythonSelf(row, 'processHash')), { spec: spec.specHash, freeze: freeze.specHash, build: build.buildHash, semantic: semantic.auditHash, receipt: receipt.receiptHash });
  gate('TOOL_BINDINGS', receipt.toolFreeze.specHash === freeze.specHash && receipt.preregistration.specHash === spec.specHash && freeze.tools.every(row => row.sha256 && row.uri), receipt.toolFreeze);
  gate('SOURCE_IMMUTABLE', receipt.source.beforeSha256 === spec.acceptedPc1Baseline.sha256 && receipt.source.afterSha256 === spec.acceptedPc1Baseline.sha256 && await shaFile(spec.acceptedPc1Baseline.sourcePath) === spec.acceptedPc1Baseline.sha256, receipt.source);
  gate('EVIDENCE_FILES', receipt.build.sha256 === await shaFile(resolve(root, 'build.json')) && receipt.build.buildHash === build.buildHash && receipt.semanticAudit.sha256 === await shaFile(resolve(root, 'semantic-audit.json')) && receipt.semanticAudit.auditHash === semantic.auditHash, { build: receipt.build, semantic: receipt.semanticAudit });
  gate('SEMANTIC_PASS', build.status === 'PASS' && semantic.status === 'PASS', { build: build.status, semantic: semantic.status });
  gate('FOUR_PHASES', JSON.stringify(semantic.phaseIds) === JSON.stringify(spec.semanticPhases.map(row => row.id).sort()) && semantic.phaseIds.length === 4, semantic.phaseIds);
  gate('FOUR_CHANNELS', build.independentActionChannels.length === spec.acceptance.minimumIndependentNonCameraActionChannels && JSON.stringify(build.independentActionChannels) === JSON.stringify(spec.independentActionChannels), build.independentActionChannels);
  gate('TARGET_SCOPE', semantic.signalTargets.length >= spec.acceptance.minimumAnimatedNonCameraTargets && semantic.signalTargets.every(name => spec.authorizedAnimatedNonCameraTargets.includes(name)), semantic.signalTargets);
  gate('SIGNAL_AMPLITUDES', semantic.signalChecks.length === build.signals.length && semantic.signalChecks.every(row => row.passed && row.peakToPeak + 1e-6 >= row.minimumPeakToPeak), semantic.signalChecks);
  gate('GEOMETRY_EXACT', JSON.stringify(build.geometryBefore) === JSON.stringify(build.geometryAfter) && JSON.stringify(semantic.geometry) === JSON.stringify(build.geometryAfter) && semantic.geometry.objects === 104 && semantic.geometry.meshes === 92 && semantic.geometry.polygons === 19810, semantic.geometry);
  gate('MATERIALS_EXACT', JSON.stringify(build.materialsBefore) === JSON.stringify(build.materialsAfter) && JSON.stringify(semantic.materials) === JSON.stringify(build.materialsAfter), semantic.materials);
  gate('PROTECTED_STATE', JSON.stringify(build.protectedStateBefore) === JSON.stringify(build.protectedStateAfter) && semantic.protectedStateCanonicalSha256 === spec.acceptedPc1Baseline.cameraLightSentinelsCanonicalSha256, semantic.protectedStateCanonicalSha256);
  gate('SHOT_STATE', JSON.stringify(build.shotsBefore) === JSON.stringify(build.shotsAfter), build.shotsAfter);
  gate('ACTION_INCREMENT', JSON.stringify(build.actionsBefore) !== JSON.stringify(build.actionsAfter) && build.addedAnimatedTargets.length >= 6 && build.addedAnimatedTargets.every(name => spec.authorizedAnimatedNonCameraTargets.includes(name)), { before: build.actionsBefore, after: build.actionsAfter, added: build.addedAnimatedTargets });
  gate('PROCESS_EXITS', processes.every(row => row.exitCode === 0 && row.peakRssBytes > 0 && row.peakRssBytes <= spec.resourceCeilings.peakRssBytesPerProcess), processes.map(row => ({ name: row.name, exitCode: row.exitCode, peakRssBytes: row.peakRssBytes })));
  let streamsPass = true; for (const row of processes) for (const stream of ['stdout', 'stderr']) streamsPass &&= row[stream].sha256 === await shaFile(resolve(repositoryRoot, row[stream].uri)) && row[stream].bytes === (await stat(resolve(repositoryRoot, row[stream].uri))).size;
  gate('PROCESS_STREAMS', streamsPass, streamsPass);
  gate('OPERATION_COUNTS', receipt.operations.BlenderStarts === 2 && receipt.operations.renderCalls === 0 && receipt.operations.derivedSceneSaves === 1 && Object.entries(receipt.operations).filter(([key]) => !['BlenderStarts', 'renderCalls', 'derivedSceneSaves'].includes(key)).every(([, value]) => value === 0), receipt.operations);
  const workFiles = await walk(WORK_ROOT), evidenceFiles = await walk(root); const workBytes = (await Promise.all(workFiles.map(stat))).reduce((sum, row) => sum + row.size, 0); const evidenceBytes = (await Promise.all(evidenceFiles.map(stat))).reduce((sum, row) => sum + row.size, 0);
  gate('RESOURCE_CEILINGS', workBytes <= spec.resourceCeilings.workBytes && evidenceBytes <= spec.resourceCeilings.evidenceBytes, { workBytes, evidenceBytes });
  gate('WORK_ROSTER_ZERO_RENDER', workFiles.some(path => path.endsWith('/PC2_ACTION_COMPLEXITY.blend')) && ![...workFiles, ...evidenceFiles].some(path => /\.(exr|png|jpg|jpeg|mov|mp4)$/i.test(path)), workFiles.map(path => relative(WORK_ROOT, path)));
  const passed = checks.filter(row => row.pass).length; if (passed !== checks.length) throw new Error(`AUDIT_${passed}_OF_${checks.length}_${checks.filter(row => !row.pass).map(row => row.id).join('_')}`);
  const audit = await writeJson(resolve(root, 'audit.json'), { schemaVersion: 'bfs.pc2IndependentAudit.v0.1', status: 'PASS', gate: 'PC.2', checks, checkPassed: passed, checkTotal: checks.length, observations: { phases: semantic.phaseIds, channels: build.independentActionChannels, signalTargets: semantic.signalTargets, signalChecks: semantic.signalChecks, geometry: semantic.geometry, workBytes, evidenceBytesBeforeAudit: evidenceBytes }, bindings: { preregistration: { uri: SPEC_URI, sha256: await shaFile(specPath), specHash: spec.specHash }, toolFreeze: { uri: FREEZE_URI, sha256: await shaFile(freezePath), specHash: freeze.specHash }, receipt: { uri: `${ROOT_URI}/receipt.json`, sha256: await shaFile(resolve(root, 'receipt.json')), receiptHash: receipt.receiptHash } }, operations: receipt.operations }, 'auditHash');
  const entries = []; for (const path of await walk(root)) { const s = await stat(path); entries.push({ path: relative(root, path), sha256: await shaFile(path), bytes: s.size }); }
  const manifest = await writeJson(resolve(root, 'root-manifest.json'), { schemaVersion: 'bfs.pc2RootManifest.v0.1', gate: 'PC.2', entries }, 'manifestHash');
  process.stdout.write(`BFS_PC2_AUDIT PASS ${passed}/${checks.length} ${audit.auditHash} ${manifest.manifestHash}\n`);
}
if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) auditPc2().catch(error => { process.stderr.write(`BFS_PC2_AUDIT_REJECTED ${error.message}\n`); process.exitCode = 1; });
